api/agent/async.go

```diff
@@ -33,17 +28,47 @@ func (a *agent) asyncDequeue() {
 			continue
 		}
 
-		// TODO output / logger should be here too...
+		callI, err := a.GetCall(FromModel(model))
+		if err != nil {
+			logrus.WithError(err).Error("error getting async call")
+			continue
+		}
 
-		a.wg.Add(1) // need to add 1 in this thread to ensure safe shutdown
+		call := callI.(*call)
+		ch := make(chan slot)
+		ctx, cancel := context.WithTimeout(ctx, 900*time.Second)
+		defer cancel()
+
+		// We'll wait up to 900 secs for resources for this async job
+		a.wg.Add(1)
 		go func() {
-			defer a.wg.Done() // can shed it after this is done, Submit will add 1 too but it's fine
+			defer a.wg.Done()
 
-			call, err := a.GetCall(FromModel(model))
+			slot, err := a.getSlot(ctx, call)
 			if err != nil {
-				logrus.WithError(err).Error("error getting async call")
-				return
+				logrus.WithFields(logrus.Fields{"id": call.Model().ID}).WithError(err).Error("error waiting for slot for async call")
+				close(ch)
+			} else {
+				ch <- slot
+			}
+		}()
+
+		// Here we wait on either shutdown or slot. If slot channel is closed, then we let this fall through,
+		// so that Submit() can attempt to get slot again and propagate the errors/logging consistently.
+		select {
+		case <-a.shutdown:
+			cancel()
+			return
+		case slot, isOpen := <-ch:
+			if isOpen {
+				// we got a slot, reserve/set it for this call to be used again by Submit() below
+				call.slot = slot
 			}
+		}
```

Comment:
this logic appears to be the same as what's in `Submit`. I'm not sure we're getting any sugar by moving it up here also, but maybe I'm missing something? If the job times out in `Submit` we should be able to handle the error that's being returned now when a job times out but hasn't been ran yet and not delete the message for it so it can be run again?

Reply:
the only issue with doing this without checking if there's _any_ ram available before trying to `Submit` or get a slot is that we're going to hold onto the message for a call for a while (likely timing out) when it could be being ran elsewhere, hence `WaitAsyncResources`.

Reply:
Yes, Call.Timeout doesn't apply the time wait in queue (or waiting for resources) basically. Except for this 900 sec wait, there's no benefit. It waits for 900 to reserve a lot. And this waiting is in one thread.

Reply:
I see, that makes sense. as alluded to in comment we're going over message timeout, I guess it seems likely that if a node hits saturation we could pretty easily run a task twice with this strategy yea?